File: scripts/regions.mjs
import { execSync } from "child_process";
import { writeFileSync } from "fs";
import { dirname, join } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const OUTPUT = join(__dirname, "../public/api/regions.csv");

/** @type {Record<string, string>} hub code → HDX org slug */
const HUBS = {
  ROSEA: "ocha-rosea",
  ROAP: "ocha-roap",
  ROWCA: "ocha-rowca",
  ROLAC: "ocha-rolac",
  ROMENA: "ocha-romena",
};

/**
 * Minimum ratio of (winner datasets / loser datasets) required to resolve a
 * conflict in favour of the dominant hub.  Below this ratio the conflict is
 * left unresolved and the country is excluded from the output.
 */
const CONFLICT_RATIO_THRESHOLD = 2;

// ---------------------------------------------------------------------------
// HTTP helper
// ---------------------------------------------------------------------------

function curlGet(/** @type {string} */ url) {
  const stdout = execSync(`curl -sf "${url}"`, { timeout: 30_000, encoding: "utf8" });
  return JSON.parse(stdout);
}

// ---------------------------------------------------------------------------
// Source 1: HDX CKAN package_search — regional hubs
//
// For each OCHA regional hub org, find all datasets they've published on HDX
// and collect the unique 3-letter country group codes attached to those datasets.
// ---------------------------------------------------------------------------

function fetchHdxHubs() {
  /** @type {Map<string, string>} iso3 → hub */
  const mapping = new Map();
  /** @type {Map<string, Set<string>>} iso3 → set of hubs (for conflict detection) */
  const allHubs = new Map();

  for (const [hub, orgName] of Object.entries(HUBS)) {
    console.log(`  HDX hubs: querying ${orgName} (${hub})...`);
    let countries;
    try {
      countries = fetchHdxOrgCountries(orgName);
    } catch (/** @type {any} */ err) {
      console.warn(`  HDX: failed for ${orgName}: ${err.message}`);
      continue;
    }
    console.log(`    ${countries.size} countries`);
    for (const iso3 of countries) {
      if (!allHubs.has(iso3)) allHubs.set(iso3, new Set());
      allHubs.get(iso3)?.add(hub);
    }
  }

  // Build final mapping; for countries in multiple hubs, resolve by dataset count
  /** @type {Map<string, { hub: string; counts: Record<string, number> } | null>} */
  const conflictResolutions = new Map();

  for (const [iso3, hubs] of allHubs) {
    if (hubs.size === 1) {
      mapping.set(iso3, [...hubs][0]);
    } else {
      // Fetch per-hub dataset counts to resolve the conflict
      /** @type {Record<string, number>} */
      const counts = {};
      for (const hub of hubs) {
        const orgSlug = HUBS[hub];
        const url =
          `https://data.humdata.org/api/3/action/package_search` +
          `?fq=organization:${orgSlug}+groups:${iso3.toLowerCase()}&rows=0`;
        try {
          const data = curlGet(url);
          counts[hub] = data.result?.count ?? 0;
        } catch {
          counts[hub] = 0;
        }
      }
      const sorted = Object.entries(counts).sort(([, a], [, b]) => b - a);
      const [winnerHub, winnerCount] = sorted[0];
      const [, loserCount] = sorted[1];
      if (loserCount === 0 || winnerCount / loserCount >= CONFLICT_RATIO_THRESHOLD) {
        mapping.set(iso3, winnerHub);
        conflictResolutions.set(iso3, { hub: winnerHub, counts });
      } else {
        conflictResolutions.set(iso3, null); // unresolved
      }
    }
  }

  return { mapping, allHubs, conflictResolutions };
}

/** @returns {Set<string>} */
function fetchHdxOrgCountries(/** @type {string} */ orgName) {
  const url =
    `https://data.humdata.org/api/3/action/package_search` +
    `?fq=organization:${encodeURIComponent(orgName)}&rows=1000&fl=groups`;
  const data = curlGet(url);
  if (!data.success) throw new Error("CKAN returned success=false");

  const countries = new Set();
  for (const pkg of data.result?.results ?? []) {
    for (const grp of pkg.groups ?? []) {
      const name = typeof grp === "string" ? grp : (grp?.name ?? "");
      if (name.length === 3 && /^[a-z]+$/i.test(name)) {
        countries.add(name.toUpperCase());
      }
    }
  }
  return countries;
}

// ---------------------------------------------------------------------------
// Analysis
// ---------------------------------------------------------------------------

/**
 * @param {Map<string, Set<string>>} allHubs
 * @param {Map<string, { hub: string; counts: Record<string, number> } | null>} conflictResolutions
 */
function printAnalysis(allHubs, conflictResolutions) {
  console.log("\n=== HDX CKAN (hub → country datasets) ===");
  for (const [hub] of Object.entries(HUBS)) {
    const countries = [...allHubs.entries()]
      .filter(([, hubs]) => hubs.has(hub))
      .map(([iso3]) => iso3)
      .sort();
    console.log(`  ${hub}: ${countries.join(" ")} (${countries.length})`);
  }

  if (conflictResolutions.size) {
    console.log(`\n  Conflict resolution (country claimed by multiple hubs):`);
    for (const [iso3, resolution] of conflictResolutions) {
      const hubs = allHubs.get(iso3) ?? new Set();
      if (resolution) {
        const countsStr = Object.entries(resolution.counts)
          .map(([h, n]) => `${h}:${n}`)
          .join(", ");
        console.log(
          `    ${iso3}: ${[...hubs].join(" + ")} → assigned to ${resolution.hub} (${countsStr})`,
        );
      } else {
        console.log(
          `    ${iso3}: ${[...hubs].join(" + ")} → unresolved (too close), excluded from output`,
        );
      }
    }
  }
}

// ---------------------------------------------------------------------------
// Write CSV
// ---------------------------------------------------------------------------

/** @param {Map<string, string>} hubMapping */
function writeCsv(hubMapping) {
  const rows = [...hubMapping.keys()].sort();
  const csv =
    "iso3,regional\n" + rows.map((iso3) => `${iso3},${hubMapping.get(iso3)}`).join("\n") + "\n";
  writeFileSync(OUTPUT, csv, "utf8");
  console.log(`\nWrote ${rows.length} rows → ${OUTPUT}`);
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

console.log("Fetching HDX CKAN hub data...");
const { mapping: hdxMapping, allHubs, conflictResolutions } = fetchHdxHubs();

printAnalysis(allHubs, conflictResolutions);

writeCsv(hdxMapping);

File: scripts/regions.py
# /// script
# requires-python = ">=3.12"
# dependencies = ["requests"]
# ///

import csv
import sys
from pathlib import Path

import requests

OUTPUT_DIR = Path(__file__).parent.parent / "public" / "api"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
OUTPUT = OUTPUT_DIR / "regions.csv"

# hub code → HDX org slug
HUBS: dict[str, str] = {
    "ROSEA": "ocha-rosea",
    "ROAP": "ocha-roap",
    "ROWCA": "ocha-rowca",
    "ROLAC": "ocha-rolac",
    "ROMENA": "ocha-romena",
}

# Minimum ratio of (winner datasets / loser datasets) required to resolve a
# conflict in favour of the dominant hub. Below this ratio the conflict is
# left unresolved and the country is excluded from the output.
CONFLICT_RATIO_THRESHOLD = 2

HDX_BASE = "https://data.humdata.org/api/3/action/package_search"


def ckan_get(url: str, params: dict) -> dict:
    res = requests.get(url, params=params, timeout=30)
    res.raise_for_status()
    return res.json()


def fetch_hdx_org_countries(org_name: str) -> set[str]:
    data = ckan_get(HDX_BASE, {"fq": f"organization:{org_name}", "rows": 1000, "fl": "groups"})
    if not data.get("success"):
        raise RuntimeError("CKAN returned success=false")
    countries: set[str] = set()
    for pkg in data.get("result", {}).get("results", []):
        for grp in pkg.get("groups", []):
            name = grp if isinstance(grp, str) else grp.get("name", "")
            if len(name) == 3 and name.isalpha():
                countries.add(name.upper())
    return countries


def fetch_hdx_hubs():
    mapping: dict[str, str] = {}
    all_hubs: dict[str, set[str]] = {}

    for hub, org_name in HUBS.items():
        print(f"  HDX hubs: querying {org_name} ({hub})...")
        try:
            countries = fetch_hdx_org_countries(org_name)
        except Exception as err:
            print(f"  HDX: failed for {org_name}: {err}", file=sys.stderr)
            continue
        print(f"    {len(countries)} countries")
        for iso3 in countries:
            all_hubs.setdefault(iso3, set()).add(hub)

    conflict_resolutions: dict[str, dict | None] = {}

    for iso3, hubs in all_hubs.items():
        if len(hubs) == 1:
            mapping[iso3] = next(iter(hubs))
        else:
            counts: dict[str, int] = {}
            for hub in hubs:
                org_slug = HUBS[hub]
                try:
                    data = ckan_get(
                        HDX_BASE,
                        {"fq": f"organization:{org_slug} groups:{iso3.lower()}", "rows": 0},
                    )
                    counts[hub] = data.get("result", {}).get("count", 0)
                except Exception:
                    counts[hub] = 0
            sorted_counts = sorted(counts.items(), key=lambda x: x[1], reverse=True)
            winner_hub, winner_count = sorted_counts[0]
            _, loser_count = sorted_counts[1]
            if loser_count == 0 or winner_count / loser_count >= CONFLICT_RATIO_THRESHOLD:
                mapping[iso3] = winner_hub
                conflict_resolutions[iso3] = {"hub": winner_hub, "counts": counts}
            else:
                conflict_resolutions[iso3] = None  # unresolved

    return mapping, all_hubs, conflict_resolutions


def print_analysis(all_hubs, conflict_resolutions):
    print("\n=== HDX CKAN (hub → country datasets) ===")
    for hub in HUBS:
        countries = sorted(iso3 for iso3, hubs in all_hubs.items() if hub in hubs)
        print(f"  {hub}: {' '.join(countries)} ({len(countries)})")

    if conflict_resolutions:
        print("\n  Conflict resolution (country claimed by multiple hubs):")
        for iso3, resolution in conflict_resolutions.items():
            hubs = all_hubs.get(iso3, set())
            if resolution:
                counts_str = ", ".join(
                    f"{h}:{n}" for h, n in resolution["counts"].items()
                )
                print(
                    f"    {iso3}: {' + '.join(hubs)} → assigned to {resolution['hub']} ({counts_str})"
                )
            else:
                print(
                    f"    {iso3}: {' + '.join(hubs)} → unresolved (too close), excluded from output"
                )


print("Fetching HDX CKAN hub data...")
hub_mapping, all_hubs, conflict_resolutions = fetch_hdx_hubs()

print_analysis(all_hubs, conflict_resolutions)

rows = sorted(hub_mapping.keys())
with open(OUTPUT, "w", newline="", encoding="utf-8") as f:
    writer = csv.writer(f)
    writer.writerow(["iso3", "regional"])
    for iso3 in rows:
        writer.writerow([iso3, hub_mapping[iso3]])

print(f"\nWrote {len(rows)} rows → {OUTPUT}")
